File: leetcode/2080.区间内查询数字的频率.py
class RangeFreqQuery:

    def __init__(self, arr: List[int]):
        self.dct = defaultdict(list)
        for i, v in enumerate(arr): self.dct[v].append(i)

    def query(self, left: int, right: int, value: int) -> int:
        return bisect.bisect_right(self.dct[value], right) - bisect.bisect_left(self.dct[value], left)

File: leetcode/2080.区间内查询数字的频率.go
type RangeFreqQuery struct {
    pos [1e4 + 1]sort.IntSlice
}


func Constructor(arr []int) RangeFreqQuery {
    q := RangeFreqQuery{}
    for i, value := range arr {
        q.pos[value] = append(q.pos[value], i)
    }
    return q
}


func (q *RangeFreqQuery) Query(left int, right int, value int) int {
    p := q.pos[value]
    return p.Search(right + 1) - p.Search(left)
}


/**
 * Your RangeFreqQuery object will be instantiated and called as such:
 * obj := Constructor(arr);
 * param_1 := obj.Query(left,right,value);
 */